Sources/CollectionDiffing.swift

```diff
@@ -0,0 +1,560 @@
+import Foundation
+import Result
+
+/// Represents a snapshot of a collection.
+///
+/// The changeset associated with the first snapshot received by any given observation
+/// is generally irrelevant. Observations should special case the first snapshot as a
+/// complete replacement, or whatever semantic that fits their purpose.
+///
+/// `Snapshot` does not include a reference to the previous snapshot. If the
+/// history is relevant for delta consumption, it is up to the observers to cache the
+/// history on their own.
+///
+/// While the protocol does not bind the `Changeset` associated type with any constraint,
+/// it is expected to be either a `Changeset`, or a composite of `Changeset`s for a nested
+/// collection.
```

Comment:
I recognize that you want this to be generic so that you can add support for nested collections, but I think it'd be better to remove this type parameter for the time being. I think it's usually best to wait to add something until you actually need it.